libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Combinators.hs

```diff
@@ -0,0 +1,206 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+
+module Test.Cardano.Ledger.Constrained.Combinators where
+
+import Cardano.Ledger.Coin (Coin (..))
+import qualified Data.List as List
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Set (Set)
+import qualified Data.Set as Set
+import GHC.Stack (HasCallStack)
+import Test.Cardano.Ledger.Core.Arbitrary ()
+import Test.QuickCheck hiding (Fixed, total)
+
+-- ==========================================================================
+-- Tracking Gen-time errors
+
+-- | Report a Gen-time error from the current message 'extra' and the
+--   [messages] 'mess' describing the path to this call site.
+errorMess :: HasCallStack => String -> [String] -> a
+errorMess extra mess = error (unlines ("\nGen-time error" : (reverse (extra : mess))))
+
+-- | suchThat version that tracks Gen-time errors
+suchThatErr :: [String] -> Gen a -> (a -> Bool) -> Gen a
+suchThatErr msgs gen p = sized $ \n -> try (10 :: Int) n
+  where
+    try 0 _ = errorMess "SuchThat times out" msgs
+    try k sz = do
+      x <- resize sz $ suchThatMaybe gen p
+      case x of
+        Just y -> pure y
+        Nothing -> try (k - 1) (sz + 5)
+
+-- =======================================================================
+
+-- | add items from 'source' to 'base' until size 'n' is reached.
+addUntilSize :: Ord a => [String] -> Set a -> Set a -> Int -> Gen (Set a)
+addUntilSize msgs base source n = do
+  let possible = Set.difference source base
+      p = Set.size possible
+      m = Set.size base
+      loop result _ | Set.size result >= n = pure result
+      loop _ extra
+        | Set.null extra =
+            errorMess
+              ( "There are not enough unused elements in 'source'("
+                  ++ show p
+                  ++ ") to reach the size 'n'("
+                  ++ show n
+                  ++ ")"
+              )
+              msgs
+      loop result extra = do
+        i <- choose (0, Set.size extra - 1)
+        loop (Set.insert (Set.elemAt i extra) result) (Set.deleteAt i extra)
+  case compare m n of
+    EQ -> pure base
+    GT -> errorMess ("The size(" ++ show m ++ ") of the 'base' set exceeds the target size(" ++ show n ++ ")") msgs
+    LT -> loop base possible
+
+setSized :: Ord a => [String] -> Int -> Gen a -> Gen (Set a)
+setSized mess size gen = do
+  set <- (Set.fromList <$> vectorOf size gen)
+  fixSet (("setSized " ++ show size) : mess) 1000 size gen set
+
+fixSet :: Ord a => [String] -> Int -> Int -> Gen a -> Set a -> Gen (Set a)
+fixSet mess numtrys size genA s = help numtrys s
```

Comment:
Left it the way it is.  I didn't write this. And don't wan to be responsible for breaking something that works.

Reply:
OK I used your code with a slight revision (scaling by 5 rather than 1). But I left the old code in a comment because when this breaks, I wan't to compare it with the previous version.

Reply:
Sounds good. The only reason why I changed to `(+1)` instead of `(+5)`, because it will be called a lot more often, then in the previous implementation. Reason for this is generation of each individual element one at a time, rather than in bulk using lists. Because it will be called more often it will scale up quicker too. At the end though, size is empirically discovered value, so it can be changed later if there are problems.